common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+// Copyright 2017-2022 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file contains generic functions and trait classes for operating on tree
+// structures.
+//
+// TreeNode concept
+// ----------------
+//
+// Class `T` must contain at least the following member in order to fulfill the
+// TreeNode concept:
+//
+// - `Container<T>& Children()`:
+//   Returns reference to a STL-like container or a range containing the node's
+//   children.
+//   The Container must at least support `begin()` and `end()` methods.
+//   Pointer and iterator stability is not required.
+//
+// To check whether class `T` fulfills the concept one can use the
+// `TreeNodeTraits<T>` traits struct. It is defined only for classes fulfilling
+// the TreeNode concept, what makes it suitable for SFINAE tests. For
+// readability and consistency with Value and Parent checks (described below)
+// there is a member `TreeNodeTraits<T>::available` which is always true.
+//
+// Optional members:
+//
+// - `T* Parent()`:
+//   Returns pointer to a parent node or nullptr when the node is a tree root.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `ValueType& Value()`:
+//   Returns reference to a value stored in the node.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `subnodes_type` (typename):
+//   Container type which should be used for storing arrays of detached
+//   child nodes. It must be move-assignable to, and move-constructible from,
+//   a value returned by `Children()` method.
+//   When the type is absent, it is assumed to be the type returned by
+//   `Children()` method with reference and other modifiers removed.
+//   It is available through `TreeNodeTraits<T>::Children::container_type`.
+
+#ifndef VERIBLE_COMMON_UTIL_TREE_OPERATIONS_H_
+#define VERIBLE_COMMON_UTIL_TREE_OPERATIONS_H_
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iosfwd>  // IWYU pragma: keep
+#include <iterator>
+#include <numeric>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "common/util/logging.h"
+#include "common/util/spacer.h"
+#include "common/util/type_traits.h"
+
+namespace verible {
+
+namespace tree_operations_internal {
+
+// TODO(mglb): move (Unavailable)FeatureTraits to a more generic header, or
```

Comment:
So far there is
https://cs.opensource.google/verible/verible/+/master:common/util/type_traits.h

Reply:
Moved it there.